File: liverrisk-webapp/frontend/predict.js
// ====================================================================
// SECTION: "Predict" tab -- DOM references
// ====================================================================
// Grab references to the page elements we need, once, at the top.
const uploadForm = document.querySelector("#upload-form");     // the <form> wrapping the file input + submit button
const fileInput = document.querySelector("#csv-file");         // the real (visually hidden) <input type="file">
const predictButton = document.querySelector("#predict-button"); // submit button, disabled while a request is in flight
const statusMessage = document.querySelector("#status-message"); // "Predicting...", "Done.", or an error line
const resultBox = document.querySelector("#result-box");       // container script.js fills with one .result-card per patient
const fileNameLabel = document.querySelector("#file-name");    // fake "chosen file" text shown over the real input


// ====================================================================
// SECTION: "Predict" tab -- file input display
// ====================================================================
// Show the chosen file's name next to the (invisible) native input.
fileInput.addEventListener("change", function () {
    const chosenFile = fileInput.files[0];
    // Fall back to the placeholder text if the user cleared the picker
    fileNameLabel.textContent = chosenFile ? chosenFile.name : "No file chosen";
    // Toggles the "chosen" class, which just changes the label's color/weight
    fileNameLabel.classList.toggle("chosen", Boolean(chosenFile));
});


// ====================================================================
// SECTION: "Predict" tab -- form submit / POST /predict
// ====================================================================
// Runs every time the user submits the form (clicks "Predict").
uploadForm.addEventListener("submit", async function (event) {
    // Stop the browser's default behavior, which would reload the page.
    event.preventDefault();

    // Make sure a file was actually chosen.
    const chosenFile = fileInput.files[0];
    if (!chosenFile) {
        statusMessage.textContent = "Please choose a CSV file first.";
        return;
    }

    // Show a loading message and disable the button so the user can't
    // click it twice while we're waiting for the server.
    statusMessage.textContent = "Predicting...";
    statusMessage.className = "";       // clear any leftover error/success styling
    resultBox.innerHTML = "";           // clear any previous result cards
    predictButton.disabled = true;

    // FormData is the standard way to send a file to a server with fetch().
    const formData = new FormData();
    formData.append("file", chosenFile);

    try {
        const response = await fetch("/predict", {
            method: "POST",
            body: formData,
        });

        // The backend sends a JSON body either way (result on success,
        // {"detail": "..."} on error), so we can parse it either way.
        const data = await response.json();

        if (!response.ok) {
            // FastAPI puts error messages in a "detail" field.
            statusMessage.textContent = "Error: " + data.detail;
            statusMessage.className = "status-error";
        } else {
            statusMessage.textContent = "Done.";
            statusMessage.className = "status-success";
            renderResults(data);
        }
    } catch (error) {
        // This happens if the server can't be reached at all.
        statusMessage.textContent = "Could not reach the server: " + error.message;
        statusMessage.className = "status-error";
    }

    // Always re-enable the button, whether the request succeeded or failed.
    predictButton.disabled = false;
});


// ====================================================================
// SECTION: "Predict" tab -- styled result card
// ====================================================================
// Builds a plain-div bar histogram from the backend's pre-binned
// {bin_start, bin_end, count} array. Bar heights are set as a % of the
// tallest bin (inline style, computed here -- not a charting library).
// The one bar the patient's own score falls into is highlighted.
function buildHistogram(bins, value) {
    const container = document.createElement("div");
    container.className = "histogram";

    const maxCount = Math.max.apply(null, bins.map(function (bin) { return bin.count; }));

    bins.forEach(function (bin, i) {
        const bar = document.createElement("div");
        bar.className = "hist-bar";
        bar.style.height = (maxCount > 0 ? (bin.count / maxCount) * 100 : 0) + "%";

        // The last bin's upper edge is inclusive, so a patient scoring
        // exactly at the training cohort's max still lands in a bar.
        const isLastBin = i === bins.length - 1;
        const inThisBin = value !== null
            && value >= bin.bin_start
            && (isLastBin ? value <= bin.bin_end : value < bin.bin_end);
        if (inThisBin) {
            bar.classList.add("hist-bar-active");
        }

        container.appendChild(bar);
    });

    return container;
}

// One "Hepatic event risk" / "Death risk" sub-section: a percentile
// headline, the underlying blended score in small print, and a
// histogram marking where this patient falls in the training cohort.
// `note` (nullable) explains a flat-looking histogram, e.g. when that
// endpoint's blend is really just one model's rank-percentile alone --
// see models_loader.py's single_model_note() -- so it doesn't read as broken.
function buildRiskSubsection(title, rawScore, percentile, histogramBins, note) {
    const section = document.createElement("div");
    section.className = "risk-subsection";

    const heading = document.createElement("h3");
    heading.textContent = title;

    const percentileLine = document.createElement("div");
    percentileLine.className = "risk-percentile";
    percentileLine.textContent = ordinal(percentile) + " percentile";

    const rawScoreLine = document.createElement("div");
    rawScoreLine.className = "risk-raw-score";
    rawScoreLine.textContent = "Blended risk score: " + rawScore;

    section.appendChild(heading);
    section.appendChild(percentileLine);
    section.appendChild(rawScoreLine);
    section.appendChild(buildHistogram(histogramBins, rawScore));

    if (note) {
        const noteLine = document.createElement("p");
        noteLine.className = "histogram-note";
        noteLine.textContent = note;
        section.appendChild(noteLine);
    }

    return section;
}

// One column ("ML blend" / "FIB-4" / "APRI") in the clinical-formula
// comparison. `percentile` is null when the patient is missing the
// labs a formula needs -- shown as plain text, not left blank.
function buildFormulaColumn(name, percentile) {
    const col = document.createElement("div");
    col.className = "formula-col";

    const label = document.createElement("div");
    label.className = "formula-name";
    label.textContent = name;

    const value = document.createElement("div");
    if (percentile === null) {
        value.className = "formula-value unavailable";
        value.textContent = "Not available (missing lab values)";
    } else {
        value.className = "formula-value";
        value.textContent = ordinal(percentile) + " percentile";
    }

    col.appendChild(label);
    col.appendChild(value);
    return col;
}

// "Why this score" section: a plain-language sentence naming this
// patient's top SHAP-ranked features (result.risk_explanation, computed
// live per patient in explain.py). Falls back to a plain unavailable
// line instead of an empty section when the backend couldn't compute one.
function buildRiskExplanation(result) {
    const section = document.createElement("div");
    section.className = "risk-explanation";

    const heading = document.createElement("h3");
    heading.textContent = "Why this score";
    section.appendChild(heading);

    const text = document.createElement("p");
    if (result.risk_explanation) {
        text.className = "risk-explanation-text";
        text.textContent = result.risk_explanation;
    } else {
        text.className = "risk-explanation-text unavailable";
        text.textContent = "Explanation not available for this patient.";
    }
    section.appendChild(text);

    return section;
}

// Row that pushes this patient's already-computed scores into the
// in-memory activePatients pool (see state.js), for the Rankings tab's
// "My active patients" scope. Uses a plain inline text input for the
// label, defaulting to "Patient N" if left blank -- NOT
// window.prompt()/confirm(), which some embedded browser contexts
// (e.g. VS Code's Simple Browser) silently block, making a
// prompt()-based button look like it does nothing at all.
function buildAddToPoolRow(result) {
    const row = document.createElement("div");
    row.className = "add-to-pool-row";

    const labelInput = document.createElement("input");
    labelInput.type = "text";
    labelInput.className = "add-to-pool-input";
    labelInput.placeholder = "Patient " + (activePatients.length + 1);
    labelInput.setAttribute("aria-label", "Label for this patient in your active pool");

    const addButton = document.createElement("button");
    addButton.type = "button";
    addButton.className = "add-to-pool-button";
    addButton.textContent = "+ Add to active pool";

    addButton.addEventListener("click", function () {
        const enteredLabel = labelInput.value.trim();
        // Computed at click time, not build time -- several cards can sit
        // unadded together (a multi-patient CSV), so activePatients.length
        // at render time would give every one of them the same "Patient 1"
        // fallback instead of counting up as they're actually added.
        const defaultLabel = "Patient " + (activePatients.length + 1);

        activePatients.push({
            label: enteredLabel === "" ? defaultLabel : enteredLabel,
            risk_hepatic_event: result.risk_hepatic_event,
            risk_death: result.risk_death,
            weighted_risk: result.weighted_risk,
            fib4_score: result.fib4_score,
            apri_score: result.apri_score,
            age_at_baseline: result.age_at_baseline,
            visit_count: result.visit_count,
        });

        addButton.textContent = "Added ✓";
        addButton.disabled = true;
        labelInput.disabled = true;
    });

    row.appendChild(labelInput);
    row.appendChild(addButton);
    return row;
}

// Builds one full .result-card for one patient's /predict result.
// `index`/`total` are only used to label the card "Patient N of M"
// when the uploaded CSV had more than one patient.
function buildResultCard(result, index, total) {
    const card = document.createElement("div");
    card.className = "result-card";

    if (total > 1) {
        const cardTitle = document.createElement("div");
        cardTitle.className = "result-card-title";
        cardTitle.textContent = "Patient " + (index + 1) + " of " + total;
        card.appendChild(cardTitle);
    }

    // The single headline figure: overall (weighted) risk percentile.
    const weighted = document.createElement("div");
    weighted.className = "weighted-risk";

    const weightedNumber = document.createElement("div");
    weightedNumber.className = "weighted-risk-number";
    weightedNumber.textContent = ordinal(result.weighted_percentile);

    const weightedLabel = document.createElement("div");
    weightedLabel.className = "weighted-risk-label";
    weightedLabel.textContent =
        "This patient ranks in the " + ordinal(result.weighted_percentile) +
        " percentile for overall risk (a 70% hepatic-event / 30% death blend), compared to the training cohort.";

    weighted.appendChild(weightedNumber);
    weighted.appendChild(weightedLabel);
    card.appendChild(weighted);

    // Hepatic event / death risk, side by side, each with its own histogram.
    const subsections = document.createElement("div");
    subsections.className = "risk-subsections";
    subsections.appendChild(buildRiskSubsection(
        "Hepatic event risk", result.risk_hepatic_event, result.hepatic_percentile,
        result.histograms.hepatic, result.distribution_notes.hepatic
    ));
    subsections.appendChild(buildRiskSubsection(
        "Death risk", result.risk_death, result.death_percentile,
        result.histograms.death, result.distribution_notes.death
    ));
    card.appendChild(subsections);

    // Hepatic-risk percentile from the ML blend next to the two
    // formula-based scores, so they can be read side by side.
    const comparison = document.createElement("div");
    comparison.className = "formula-comparison";

    const comparisonHeading = document.createElement("h3");
    comparisonHeading.textContent = "Compared to clinical formulas";
    comparison.appendChild(comparisonHeading);

    const columns = document.createElement("div");
    columns.className = "formula-columns";
    columns.appendChild(buildFormulaColumn("ML blend (hepatic)", result.hepatic_percentile));
    columns.appendChild(buildFormulaColumn("FIB-4", result.fib4_percentile));
    columns.appendChild(buildFormulaColumn("APRI", result.apri_percentile));
    comparison.appendChild(columns);
    card.appendChild(comparison);

    card.appendChild(buildRiskExplanation(result));

    card.appendChild(buildAddToPoolRow(result));

    return card;
}

// Renders the full /predict response (a list of per-patient results)
// as one result card per patient, replacing whatever was in the box.
function renderResults(results) {
    resultBox.innerHTML = "";
    results.forEach(function (result, i) {
        resultBox.appendChild(buildResultCard(result, i, results.length));
    });
}

File: liverrisk-webapp/backend/predict.py
"""
LiverRisk webapp backend -- POST /predict route.

Accepts a CSV upload (one or a few patients, same raw columns as
train.csv / test.csv), runs them through the already-trained models
loaded in models_loader.py, and returns a risk score for each patient.
"""
from __future__ import annotations

import io

import pandas as pd
from fastapi import APIRouter, File, HTTPException, UploadFile

# models_loader must be imported before any liverrisk.* import below --
# it's the one that puts the repo root (where the liverrisk package
# lives) onto sys.path.
from models_loader import (
    AGE_COLUMNS,
    DISTRIBUTION_NOTES,
    FEATURE_COLUMNS,
    HISTOGRAMS,
    REQUIRED_COLUMNS,
    TRAIN_FIB4_SCORES,
    TRAIN_APRI_SCORES,
    TRAIN_WEIGHTED_SCORES,
    X_HEP,
    _json_safe,
    percentile_le,
    score_one_endpoint,
)

from explain import explain_patient

from liverrisk.clinical_scores import compute_fib4_apri
from liverrisk.features import build_patient_features

router = APIRouter()


# --------------------------------------------------------------------
# The one API endpoint
# --------------------------------------------------------------------
@router.post("/predict")
async def predict(file: UploadFile = File(...)):
    """
    Accepts a CSV upload (one or a few patients, same columns as
    train.csv/test.csv) and returns a risk score + percentile for
    each patient, for both the hepatic-event and death outcomes.
    """
    # Step 1: read the upload into a DataFrame.
    raw_bytes = await file.read()
    try:
        raw_df = pd.read_csv(io.BytesIO(raw_bytes))
    except Exception:
        raise HTTPException(status_code=400, detail="Could not read the uploaded file as a CSV.")

    # Step 2: check the CSV has the columns we need before we try
    # anything fancier, so the error message is clear.
    missing_columns = [c for c in REQUIRED_COLUMNS if c not in raw_df.columns]
    if missing_columns:
        raise HTTPException(
            status_code=400,
            detail=f"The uploaded CSV is missing required columns: {', '.join(missing_columns)}",
        )

    # Step 3: build the same engineered features the models were
    # trained on, then line up the columns in the exact order the
    # models expect.
    try:
        patient_features = build_patient_features(raw_df)
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Could not build features from the uploaded CSV: {e}")

    patient_features = patient_features.reindex(columns=FEATURE_COLUMNS)

    # A fresh upload can infer different dtypes than training did (e.g.
    # int64 for a column that happens to have no missing values in this
    # particular upload, vs. float64 during training where the same
    # column had NaNs). Casting to X_HEP's dtypes keeps every column on
    # the exact dtype the fitted preprocessors expect, before any model
    # ever sees this data.
    patient_features = patient_features.astype(X_HEP.dtypes.to_dict())

    # FIB-4/APRI for every uploaded patient at once (vectorized), using
    # the same clinical_scores.compute_fib4_apri() helper the training-
    # cohort reference distribution was built from above.
    patient_fib4_apri = compute_fib4_apri(patient_features, "hep")

    # Step 4: score every patient (row) in the upload.
    results = []
    for i in range(len(patient_features)):
        one_patient = patient_features.iloc[[i]]  # keep it as a 1-row DataFrame

        risk_hepatic_event, hepatic_percentile = score_one_endpoint(one_patient, "hep")
        risk_death, death_percentile = score_one_endpoint(one_patient, "death")
        risk_explanation = explain_patient(one_patient)

        # Same formula build_submission() (scripts/train.py) uses to turn
        # a test-set patient's two blended scores into one weighted_risk.
        weighted_risk = round(0.7 * risk_hepatic_event + 0.3 * risk_death, 4)
        weighted_percentile = percentile_le(weighted_risk, TRAIN_WEIGHTED_SCORES)

        fib4_score = patient_fib4_apri["fib4_score"].iloc[i]
        apri_score = patient_fib4_apri["apri_score"].iloc[i]
        fib4_score = None if pd.isna(fib4_score) else round(float(fib4_score), 2)
        apri_score = None if pd.isna(apri_score) else round(float(apri_score), 2)
        fib4_percentile = None if fib4_score is None else percentile_le(fib4_score, TRAIN_FIB4_SCORES)
        apri_percentile = None if apri_score is None else percentile_le(apri_score, TRAIN_APRI_SCORES)

        # Age_v1 is guaranteed present (it's in REQUIRED_COLUMNS); the
        # rest of AGE_COLUMNS (Age_v2, Age_v3, ...) may not all be in a
        # smaller upload, so only count the ones actually present --
        # same visit-counting logic as select_sample_patients() in
        # sample_patients.py, just applied to whatever columns this
        # particular CSV has.
        # Used by the Rankings tab's "My active patients" scope, which
        # gets a patient's age/visits from this response rather than
        # re-parsing the CSV itself.
        present_age_columns = [c for c in AGE_COLUMNS if c in raw_df.columns]
        age_at_baseline = _json_safe(raw_df.iloc[i]["Age_v1"])
        visit_count = int(raw_df.iloc[i][present_age_columns].notna().sum())

        results.append({
            "risk_hepatic_event": risk_hepatic_event,
            "hepatic_percentile": hepatic_percentile,
            "risk_death": risk_death,
            "death_percentile": death_percentile,
            "risk_explanation": risk_explanation,
            "weighted_risk": weighted_risk,
            "weighted_percentile": weighted_percentile,
            "fib4_score": fib4_score,
            "fib4_percentile": fib4_percentile,
            "apri_score": apri_score,
            "apri_percentile": apri_percentile,
            "age_at_baseline": age_at_baseline,
            "visit_count": visit_count,
            "histograms": HISTOGRAMS,
            "distribution_notes": DISTRIBUTION_NOTES,
        })

    return results
